cpp/src/genetic/program.cu

```diff
@@ -0,0 +1,565 @@
+                 const int n_rows,
+                 const float* data,
+                 const float* y,
+                 const float* sample_weights)
+{
+  cudaStream_t stream = h.get_stream();
+
+  rmm::device_uvector<float> score(1, stream);
+
+  find_fitness(h, d_prog, score.data(), params, n_rows, data, y, sample_weights);
+
+  // Update host and device score for program
+  CUDA_CHECK(cudaMemcpyAsync(
+    &d_prog[0].raw_fitness_, score.data(), sizeof(float), cudaMemcpyDeviceToDevice, stream));
+  h_prog.raw_fitness_ = score.front_element(stream);
+}
+
+void set_batched_fitness(const raft::handle_t& h,
+                         int n_progs,
+                         program_t& d_progs,
+                         std::vector<program>& h_progs,
+                         const param& params,
+                         const int n_rows,
+                         const float* data,
+                         const float* y,
+                         const float* sample_weights)
+{
+  cudaStream_t stream = h.get_stream();
+
+  rmm::device_uvector<float> score(n_progs, stream);
+
+  find_batched_fitness(h, n_progs, d_progs, score.data(), params, n_rows, data, y, sample_weights);
+
+  // Update scores on host and device
+  // TODO: Find a way to reduce the number of implicit memory transfers
+  for (auto i = 0; i < n_progs; ++i) {
+    CUDA_CHECK(cudaMemcpyAsync(&d_progs[i].raw_fitness_,
+                               score.element_ptr(i),
+                               sizeof(float),
+                               cudaMemcpyDeviceToDevice,
+                               stream));
+    h_progs[i].raw_fitness_ = score.element(i, stream);
+  }
+}
+
+float get_fitness(const program& prog, const param& params)
+{
+  int crit      = params.criterion();
+  float penalty = params.parsimony_coefficient * prog.len * (2 * crit - 1);
+  return (prog.raw_fitness_ - penalty);
+}
+
+/**
+ * @brief Get a random subtree of the current program nodes (on CPU)
+ *
+ * @param pnodes  AST represented as a list of nodes
+ * @param len     The total number of nodes in the AST
+ * @param rng     Random number generator for subtree selection
+ * @return A tuple [first,last) which contains the required subtree
+ */
+std::pair<int, int> get_subtree(node* pnodes, int len, std::mt19937& rng)
+{
+  int start, end;
+  start = end = 0;
+
+  // Specify RNG
+  std::uniform_real_distribution<float> dist_uniform(0.0f, 1.0f);
+  float bound = dist_uniform(rng);
+
+  // Specify subtree start probs acc to Koza's selection approach
+  std::vector<float> node_probs(len, 0.1);
+  float sum = 0.1 * len;
+
+  for (int i = 0; i < len; ++i) {
+    if (pnodes[i].is_nonterminal()) {
+      node_probs[i] = 0.9;
+      sum += 0.8;
+    }
+  }
+
+  // Normalize vector
+  for (int i = 0; i < len; ++i) {
+    node_probs[i] /= sum;
+  }
+
+  // Compute cumulative sum
+  std::partial_sum(node_probs.begin(), node_probs.end(), node_probs.begin());
+
+  // CUML_LOG_DEBUG("Current bound is %f",bound);
```

Comment:
If we must comment this out, can we just remove it?

Reply:
yes i have removed it corey 👍🏾